### PR TITLE
FIX: prevent `support_usm_ndarray` from changing queue if explicitly provided.

### PR DESCRIPTION
**Pull request by olegkkruglov**
### Description 

* Changed logic of `support_usm_ndarray`. Now it does not change the queue if it was explicitly provided to args. Previous logic does not allow to use `np.array` + `SyclQueue` interface with decorated method.
* `queue_param` is removed from `support_usm_ndarray` since the check can be done via kwargs checking.
* Docstring is added to `support_usm_ndarray` with highlighting the change

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation to reflect the changes, if necessary
- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.
- [x] I have resolved any merge conflicts that might occur with the base branch.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_
- [x] I have added a respective label(s) to PR if I have a permission for that.

### CODE REVIEW

File: onedal/_device_offload.py [hunk 1]
```diff
@@ -178,11 +178,23 @@ def _convert_to_dpnp(array):
...
+                         Queue will not be changed if provided explicitly.
+    """
+
     def decorator(func):
         def wrapper_impl(obj, *args, **kwargs):
             usm_iface = _extract_usm_iface(*args, **kwargs)
             data_queue, hostargs, hostkwargs = _get_host_inputs(*args, **kwargs)
-            if queue_param:
+            if queue_param and not (
+                "queue" in hostkwargs and hostkwargs["queue"] is not None
```
Comment (samir-nasibli) [inline onedal/_device_offload.py:196]:
```suggestion
                "queue" in hostkwargs and hostkwargs["queue"] is None
```
queue is prior than data_queue

**Review by samir-nasibli — COMMENTED** [review]
Thank you for founding out this issue.
This is critical for sklearnex.spmd ifaces with numpy + sycl_queue. Unfortunately we don't have properly testing yet for this API. Only some examples with dpnp/dpctl input, that not the case of this bug. Hopefully development on #1777 and #1909 will cover this cases as well.

I would like to have some tests in case. I can follow this add it here.

File: onedal/_device_offload.py [hunk 2]
```diff
@@ -178,11 +178,23 @@ def _convert_to_dpnp(array):
...
 
 def support_usm_ndarray(freefunc=False, queue_param=True):
+    """
+    Handles USMArray input. Puts SYCLQueue from data to method arguments.
+
+    Parameters
+    ----------
+    freefunc (bool) : Set to True if decorates free function.
+    queue_param (bool) : Set to True if queue from data should be provided to the method args.
+                         Queue will not be changed if provided explicitly.
```
Comment (samir-nasibli) [inline onedal/_device_offload.py:188]:
```suggestion
    queue_param (bool) : Set to True if decorated function has a `queue` parameter.
                         Queue will not be changed if provided explicitly.
```

File: onedal/_device_offload.py (hunk 1 above)
Comment (samir-nasibli) [inline onedal/_device_offload.py:196]:
```suggestion
                "queue" in hostkwargs and hostkwargs["queue"] is None
```
I think provided queue param value is more prior that input data's context. But maybe some checks are required

Comment (olegkkruglov) [reply onedal/_device_offload.py:196]: I can't get your suggested change. Currently the condition in this row means that queue is provided and not equal to None, so shouldn't be changed

File: onedal/_device_offload.py (hunk 2 above)
Comment (olegkkruglov) [reply onedal/_device_offload.py:188]: I would say that my comment makes more sense. It might be possible that decorated function has `queue` parameter but user doesn't want to put queue from data there.

File: onedal/_device_offload.py (hunk 1 above)
Comment (samir-nasibli) [reply onedal/_device_offload.py:197]: Ah, sorry, didn't see not before the braces.

File: onedal/_device_offload.py [hunk 3]
```diff
@@ -178,11 +178,24 @@ def _convert_to_dpnp(array):
...
+    """
+
     def decorator(func):
         def wrapper_impl(obj, *args, **kwargs):
             usm_iface = _extract_usm_iface(*args, **kwargs)
             data_queue, hostargs, hostkwargs = _get_host_inputs(*args, **kwargs)
-            if queue_param:
+            if queue_param and not (
+                "queue" in hostkwargs and hostkwargs["queue"] is not None
+            ):
```
Comment (samir-nasibli) [inline onedal/_device_offload.py:198]:
```suggestion
            if not ("queue" in hostkwargs and hostkwargs["queue"] is not None):
```
This param were added on #1832, now it looks overkill for this decorator. Maybe it make sense just remove `queue_param` usage and param from `support_usm_ndarray`and just leave this condition checking.

Just needed to remove all `queue_param=False` from everywhere.

File: onedal/_device_offload.py (hunk 2 above)
Comment (samir-nasibli) [reply onedal/_device_offload.py]: outdated

File: onedal/_device_offload.py [hunk 4]
```diff
@@ -178,11 +178,25 @@ def _convert_to_dpnp(array):
...
 def support_usm_ndarray(freefunc=False, queue_param=True):
+    """
+    Handles USMArray input. Puts SYCLQueue from data to decorated function arguments.
+    Converts output of decorated function to dpctl.tensor/dpnp.ndarray if input was of this type.
+
+    Parameters
+    ----------
+    freefunc (bool) : Set to True if decorates free function.
+    queue_param (bool) : Set to True if queue from data should be provided to the decorated function args
+                        (recommended to set to True if decorated function has `queue` parameter).
```
Comment (samir-nasibli) [inline onedal/_device_offload.py:189]:
```suggestion
                        Note: set to True if decorated function has `queue` parameter.
```

**Review by samir-nasibli — CHANGES_REQUESTED** [review]

**Review by samir-nasibli — COMMENTED** [review]
We went a little bit wrong in our thoughts, but it seems like it should be normal now.

Again without testing this primitive we could possible have some other issues. It would be great to add tests to check for this estimator expected behavior for the dummy function. Could be done in separate PR as follow up, but make sure that we have ticket covered. Thank you!

**Review by samir-nasibli — APPROVED** [review]
LGTM. Assuming green CI

File: onedal/_device_offload.py (hunk 3 above)
Comment (samir-nasibli) [reply onedal/_device_offload.py:201]: Outdated.

**Comment by ethanglaser** [comment]
Please run black formatting on the file

**Review by ethanglaser — APPROVED** [review]

**Comment by samir-nasibli** [comment]
@mergify backport rls/2024.6.0-rls
